source/extensions/geoip_providers/maxmind/geoip_provider.cc

```diff
@@ -225,15 +271,90 @@ void GeoipProvider::lookupInAnonDb(
                                   config_->anonProxyHeader().value(), MMDB_ANON_LOOKUP_ARGS[4]);
         }
         if (lookup_result.size() > n_prev_hits) {
-          config_->incHit("anon_db");
+          config_->incHit(ANON_DB_TYPE);
         }
         MMDB_free_entry_data_list(entry_data_list);
       } else {
-        config_->incLookupError("anon_db");
+        config_->incLookupError(ANON_DB_TYPE);
       }
     }
-    config_->incTotal("anon_db");
+    config_->incTotal(ANON_DB_TYPE);
+  }
+}
+
+MaxmindDbSharedPtr GeoipProvider::initMaxmindDb(const std::string& db_path,
+                                                const absl::string_view& db_type, bool reload) {
+  MMDB_s maxmind_db;
+  int result_code = MMDB_open(db_path.c_str(), MMDB_MODE_MMAP, &maxmind_db);
+  if (reload && MMDB_SUCCESS != result_code) {
+    ENVOY_LOG(error, "Failed to reload Maxmind database {} from file {}. Error {}", db_type,
+              db_path, std::string(MMDB_strerror(result_code)));
+    return nullptr;
+  } else {
+    RELEASE_ASSERT(MMDB_SUCCESS == result_code,
+                   fmt::format("Unable to open Maxmind database file {}. Error {}", db_path,
+                               std::string(MMDB_strerror(result_code))));
+    ENVOY_LOG(info, "Succeeded to reload Maxmind database {} from file {}.", db_type, db_path);
+    return std::make_shared<MMDB_s>(maxmind_db);
+  }
+}
+
+absl::Status GeoipProvider::mmdbReload(const MaxmindDbSharedPtr reloaded_db,
+                                       const absl::string_view& db_type) {
+  if (reloaded_db) {
+    if (db_type == CITY_DB_TYPE) {
+      updateCityDb(reloaded_db);
+      config_->incDbReloadSuccess(db_type);
+    } else if (db_type == ISP_DB_TYPE) {
+      updateIspDb(reloaded_db);
+      config_->incDbReloadSuccess(db_type);
+    } else if (db_type == ANON_DB_TYPE) {
+      updateAnonDb(reloaded_db);
+      config_->incDbReloadSuccess(db_type);
+    } else {
+      ENVOY_LOG(error, "Unsupported maxmind db type {}", db_type);
+      return absl::InvalidArgumentError(fmt::format("Unsupported maxmind db type {}", db_type));
+    }
+  } else {
+    config_->incDbReloadError(db_type);
   }
+  return absl::OkStatus();
+}
+
+MaxmindDbSharedPtr GeoipProvider::getCityDb() const ABSL_LOCKS_EXCLUDED(mmdb_mutex_) {
+  absl::ReaderMutexLock lock(&mmdb_mutex_);
+  return city_db_;
+}
+
+void GeoipProvider::updateCityDb(MaxmindDbSharedPtr city_db) ABSL_LOCKS_EXCLUDED(mmdb_mutex_) {
+  absl::MutexLock lock(&mmdb_mutex_);
+  city_db_ = city_db;
```

Comment:
@nezdolik for the test failures about memory leak, is it because we need to do `MMDB_close` for the old db pointer to release? we have `MMDB_close` currently in the class destructor, which works for the non-reload case

Reply:
tested in https://github.com/envoyproxy/envoy/pull/34890 seems this would fix the ASAN tests

Reply:
thanks @hwyuan, i will update the patch today. Appreciate the help!